### ISSUE TITLE
Questions and missing Feature SMS Reminder Button and Login Autoclick

### ISSUE THREAD

**Issue by Caliban2017** [SamMackrill/gw2-alternator#40]
Hello, i found your Tool today here and it looks exactly like that what i looked for, to autocollect the daily Rewards fast and easy on multiple accounts. But i played around now a little bit and some things are a bit confusing.

My Questions:
What does the Buttons "Force All" and "Force Serial" in the top left? Does is start all Accounts at once/in Serial?

What does the Login and Collect Buttons? Seems they do both the same?

I have noticed that there are some problems with fresh Accounts. They ask for the first few logins for a Verification Code that comes via E-Mail. But the window awaiting the Code to Enter automatically closes after 10-20 Seconds. Its not possible to copy the Code from the E-Mail into the Window before it relaunches. After the relaunch there is a new Code and this will get an endless Loop. I must use Launch Body again for that to enter the Code withouth difficulties. 

And there is one Feature Request: is it possible to add a Auto-Click function for the SMS-Reminder? Every second start of the Game, it asks for a Phone Number to enter for a 2-way auth. Is it possible to add an autoclick option, that simply click the "later" button, to skip that window and proceed to login? Same for the Login Button. Launchbuddy already have that Feature, it will automatically click the Login Button.

And one last question: you say on the mainpage it will cause problems logging in >10 Accounts. Is that meaning at the same time? So when i login at 3 Accounts at a time and close the game after collecting the reward and launch the next 3, will that be affected by the Problem? What would be the savest way without VPN in that case? Login at 9 Accounts maximum, wait some hours and do the next 9 to be very safe?

Thanks for your work, hope i can give some usefull Input to improve your Tool!

Greetings from Germany

**Comment by SamMackrill** [comment]
Force All: If you have already logged-in some accounts previously that day then it will automatically skip them, this lets you override that feature. This is not often useful unless things get confused.

Force Serial: Each account now logs in one at a time (normally they are processed in parallel). This is useful if the login servers are temperamental, it is slow but reliable.

Login is for claiming the daily rewards, GW2 will auto shutdown as soon as it can
Collect is for harvesting the accounts, here GW2 will stay open until you close it

I am assuming that all your accounts have been logged in previously by hand and the SMS verification has been done and you have checked the Trust this IP so you do not get the SMS verification (repeated for each VPN as required). Perhaps I should write something on this.

ArenaNet limit the logins per second and this can affect you, the program should automatically slow down if it detects login failures but this is a bit hit or miss as it depends on the login servers and your pervious attempts. Generally the program will cope on its own and you can just leave it whilst you have some Frühstück :) I am still trying to workout exactly how arenanet throttle the logins, but it is hard to guess. You don't need to wait hours, 15 mins is generally enough. But as I say the program should do all that. Let me know how you get on.

**Comment by SamMackrill** [comment]
I have updated the readme to make things clearer, let me know any more clarifications I can make.

**Comment by Caliban2017** [comment]
Thank you so far for your Answers, this brings some Light for me on these Things =)

Some things wasn't clear at the beginning, maybe some Tooltip Help would help here. 

I have at the Moment 15 Accounts, had 6 before, it worked fine with Launch Buddy so far, no Login Errors. Now i created some Accounts thanks to the nvidia Heoric Edition Key giveaway on last weekend. Saw it sadly too late, thought it last a bit longer then the Weekend^^ But 15 is to much now to do it manually every day, so your Tool is very welcome :) 

I will try it tomorrow for the first time collecting the Daily Reward. I try now to activate the SMS Auth on all that accounts. I hope it will not make any problems using only one Phone Number for all Accs. But i hope the Reminder Window is gone after it and i can use the Tool without stops :)

But at the End: i noticed one more bug, when importing from Launch Buddy. One of the Accounts have a E-Mail Adresse in Form of "lulu-blabla@web.de" (only a example of course ^^). In the UI it shows under Account "lulu" and the Character Name says blabla@web.de in this case. So the "-" split up for some reason the Account name into some part of the Character Name. All other E-Mails does not contain a "-" and they all show the name@provider.com as Account Name and the Character says "Unknown" (no API Key entered on them).

Thank you very much for your work! Please keep doing it, this might be a real Launch Buddy Killer with some more Features and a cleaner UI with more Options =)

**Comment by SamMackrill** [comment]
FYI if you use google then you can use the same email for all your accounts (by adding a + postfix):
see here: https://support.google.com/a/users/answer/9308648?hl=en

**Comment by SamMackrill** [comment]
I added a proper bug for the email import, will tale a look tomorrow.

**Comment by Caliban2017** [comment]
I've tried some Accounts for now but i am still a little confused... I still don't understand the Reason for the two different Buttons for Login and Collect, because they do the same thing for me. I thought Login should start the Game by itself, login and choose a Char to load into the Game, then auto-collect the Daily Reward and then it should close the Game when its done an proceed with the next Accounts. And Collect is to do it manually and let the Game open after collecting the Daily Reward. But sometimes it does not log into a Char and it ends up in the Character Main Screen. When i now choose the Char manually, close the Game and try it again, it starts and this time it auto choose that Char and go into the Game. But then nothings happens... I must manually click the Daily Chest and close the Game by myself. But then it try to login again. It is not marked as Completed, it stays on unset. On the Collect Button exactly the same happens, but there is no restart after closing the Game. But it does not collect anything.

**Comment by SamMackrill** [comment]
I think the confusion is on the approach here as it is not efficient to click on the daily chests every day, there is no harm in leaving them bouncing in the corner for weeks. This is the assumed by me. I should clarify this somewhere.


Login is to do the absolute minimum to be awarded the daily reward. You do not need to click on the chest as they can stack-up and be collected in a group later (say once a month). So it should do:

Launch => Play => Enter World => Immediately close

Login is something you do every day


Collect is to let you click on the chests get laurels, transfer materials to your guild bank etc. This can take a while and should do:

Launch -> Play => Enter World => Minimise until ready => Maximise => wait for you to close GW2

Collect is something you do less regularly as it is much slower. There is a column that assumes you harvest once a month.

**Comment by SamMackrill** [comment]
I will work on documenting this better

**Comment by Caliban2017** [comment]
Ahh okay, that make sense. I didn't mind that the daily chests will be stacked. But it is still not working properly with the Login Button for me. It starts the Game but most of the time it stucks at the choose Character Screen and nothings happens. When i close the Game, Alternator is instantly restarting the Game with the same result, stuck at the Mainscreen. I must use the "Stop!" Button to stop that loop. Sometimes the Login work and i see a loading Screen and i am in the Game but then nothing happens. The Game does not close itself, it stays open. Because of that i was confused, i awaited here a automatic Input to receive the Chest or something. And of course here will happen the same, when i close the Game, Alternator instantly relaunch the Game.

And i found one more little Bug maybe: the Last Login Timestamp is not refreshing. I logged in two times with one of the Accounts but it always stays on the same Timestamp.

**Comment by SamMackrill** [comment]
![image](https://user-images.githubusercontent.com/84898/166441105-42ad9e24-a39c-4f59-8ccf-78de659dee39.png)
make sure these are set

**Comment by SamMackrill** [comment]
If you close the game then it thinks it has crashed and so tries again. The timestamp only updates on success. I am working on it better diagnosing when things go wrong but it is pretty tricky.

**Comment by Caliban2017** [comment]
Yes they are set. The Login works fine, but it stops at the Character Screen. 

I tried around a bit and noticed some things:

when i start Alternator, all Accounts Status say "Unset". When i login to one account, i stuck at the Char Screen. When i then manually choose a Char and let it load and then close the Game, it tries to instantly restart the Game but the Status now says "ready". When i then use the Login Button again on that Account, it load into the Game, but then nothing happens. 
When i close now the Alternator and restart it, all the Status are reset to "unset". And when i now try to login to the same account again, it stucks again in the Char Screen. It seems like Alternator is "forgetting" the Settings after a restart...?

**Comment by Caliban2017** [comment]
Maybe something is causing problems here? Can ArcDPS be the Problem for this? Or the OS? I am using Windows 11 x64 Pro. How is alternator detecting that the Char is logged into the Game?

**Comment by SamMackrill** [comment]
I have arcdps too. If you go here: %AppData%\Gw2-Alternator
There should be a log file and an error file
![image](https://user-images.githubusercontent.com/84898/166446767-2da8c4e1-41bf-4753-ac51-b26292887f4a.png)
You can have a look in there if you feel brave or just zip them up and email them to me: sam.mackrill [AT] gmail.com

**Comment by SamMackrill** [comment]
Sorry for the issues! Thanks for being patient and helping me fix them.

**Comment by SamMackrill** [comment]
State change is detected via memory increasing, I should add those thresholds as settings. I have tried many ways to detect this more reliably but nothing seems to work except this and it seems this needs some tuning
`        async Task CheckMemoryThresholdReached(long diff, long memoryUsage)
        {
            if (RunStage is not (RunStage.Authenticated or RunStage.CharacterSelected)) return;

            if (diff >= tuning.MinDiff) return;

            //Logger.Debug("{0} Memory={1} ({2}<{3})", Account.Name, memoryUsage, diff, tuning.MinDiff);
            switch (RunStage)
            {
                case RunStage.Authenticated when memoryUsage > 120_000:
                    await ChangeRunStage(RunStage.ReadyToPlay, 4000, "Memory threshold", cancellationToken);
                    break;
                case RunStage.CharacterSelected when memoryUsage > 1_400_000:
                    await ChangeRunStage(RunStage.WorldEntered, 1800, "Memory threshold", cancellationToken);
                    break;
            }
        }`

**Comment by Caliban2017** [comment]
I am glad when i can help you :) I zipped the Logs and send them to you by E-Mail. 

i deleted all logs and created some fresh ones. In this Case i tried to Log into Acc 1, it stuck at the Char Screen again, must closed the Window manually. Then it automatically restarted, now it loaded into the game but then nothing happened again, i must close the window manually. Now i pressed the stop button and closed the program...

**Comment by SamMackrill** [comment]
Ok investigating.

**Comment by SamMackrill** [comment]
I added some extra module checks that may solve your problem and some more logging. Try V1.0.17 and let me know

**Comment by Caliban2017** [comment]
Thank you for your work. I tried to download the v1.0.17 but Chrome instantly deleted it and says Error - Virus found. Windows Defender reporting the same: Trojan:Win32/Wacatac.B!ml ... Maybe something went wrong here? :(

**Comment by SamMackrill** [comment]
I just downloaded it, seems fine to me. Nothing I added should trigger AntiVirus
Seems like this can happen with C# programs: https://answers.microsoft.com/en-us/protect/forum/all/what-to-do-about-trojanwin32wacatacbml-false/d884df80-2dba-46c9-bd8b-4e2f93c9dd3f
I don't run Windows Defender I use Panda Dome (free version)

**Comment by Caliban2017** [comment]
Maybe that helps to identifiy the Problem? Maybe the Detection of the DLL Files you mentioned? 

![av1](https://user-images.githubusercontent.com/32013635/166587805-856f1f62-3fab-45f8-bd90-d9f2a98d5277.jpg)

**Comment by SamMackrill** [comment]
![image](https://user-images.githubusercontent.com/84898/166588426-b7df2b66-2477-4935-83df-ad6ba320d9ca.png)
Pretty sure this is a false positive. Defender is know to do this. I will try a new build to see if it is happier.

**Comment by SamMackrill** [comment]
Also the application is fully built in the cloud by github automation. So it is almost impossible that it it infected.

**Comment by Caliban2017** [comment]
I tried now this Version, online checks also say for the most no Virus found. 

I tried now two Accounts, i'm not sure if this is going correct now. It starts, i see the Character Menu for about 1 Second, then the Game closes. No loading Screen.

The Log say now this:

`2022-05-04 01:30:46.3263|DEBUG|guildwars2.tools.alternator.MVVM.model.Client|Acc 1 Change State to Playing because Module mmdevapi.dll loaded
2022-05-04 01:30:46.3263|DEBUG|guildwars2.tools.alternator.Launcher|Acc 1 Status changed from ReadyToPlay to Playing
2022-05-04 01:30:47.8362|DEBUG|guildwars2.tools.alternator.MVVM.model.Client|Acc 1 Module: mscms.dll
2022-05-04 01:30:48.0483|DEBUG|guildwars2.tools.alternator.MVVM.model.Client|Acc 1 Change State to CharacterSelectReached because Module mscms.dll loaded
2022-05-04 01:30:48.0483|DEBUG|guildwars2.tools.alternator.Launcher|Acc 1 Status changed from Playing to CharacterSelectReached
2022-05-04 01:30:48.0483|DEBUG|guildwars2.tools.alternator.MVVM.model.Client|Acc 1 Send ENTER
2022-05-04 01:30:50.2816|DEBUG|guildwars2.tools.alternator.MVVM.model.Client|Acc 1 Change State to CharacterSelected because Memory increased by 454612 > 100
2022-05-04 01:30:50.2816|DEBUG|guildwars2.tools.alternator.Launcher|Acc 1 Status changed from CharacterSelectReached to CharacterSelected
2022-05-04 01:30:51.0154|DEBUG|guildwars2.tools.alternator.MVVM.model.Client|Acc 1 Change State to Exited because Process.Exit event
2022-05-04 01:30:51.0154|DEBUG|guildwars2.tools.alternator.Launcher|Acc 1 Status changed from CharacterSelected to Exited
2022-05-04 01:30:51.0154|DEBUG|guildwars2.tools.alternator.MVVM.model.Client|Acc 1 GW2 process exited
2022-05-04 01:30:53.0168|DEBUG|guildwars2.tools.alternator.Launcher|Acc 1 exe semaphore released, count=3`

**Comment by SamMackrill** [comment]
That looks right now! 
The Status on the UI should have cycled through to complete the Last Login updated and the Logon? column changed to No.

This is how it is supposed to behave it exits as soon as it can. 
So you should be able to run all accounts in a few minutes. There is a link on the home page to a google doc and you can see my results for about 40 accounts. Generally takes about 10-15 minutes with no input from me.

**Comment by Caliban2017** [comment]
Cool, finally :) I tested it a few minutes ago, worked fine for two accounts so far! Must still enter the Verification Code until all Networks are trusted X.x 

So there is no need to log into the Charakter? You get the Daily Chest only by Login to the Main Menu?

**Comment by SamMackrill** [comment]
You do need to login to the character, but not all the way with all the map and models loaded so it is safe just to kill the process almost immediately and you still have the reward. This saves a lot of time.

**Comment by Caliban2017** [comment]
That is really fast now, i only see the Char Screen for less that a second, it instantly closes. But the UI says completed. But after a restart of Alternator, the Status of the Account resets to Unset. Wouldn't it be better to safe the state until the next Daily Reset? 

Can i test it somehow if it worked or do i must wait until tomorrow to see if there are two chests? The UI say on all Accounts 1 Chest. Does it get this Number from the API?? Then i would say its working =)

**Comment by SamMackrill** [comment]
That is what the Last Login and Login? columns are for. The state column is really just current run and may be used for other types of run e.g. updating client versions. 

Yes, unfortunately due to the nature of things you need to wait another day before checking your reward chest count. The count in the UI is just based on a successful login count. I don't think it is possible to see unclaimed chests from the API.

Did you add API keys to the accounts then? I am still working on the mystic coin/laurel estimation (sometimes seems way off)

Vielen dank für die hilfe

**Comment by Caliban2017** [comment]
Yes i noticed it, you can see on the Timestamp which Accounts already completed. This works of course too =)

Okay then i have to wait, but i report back of course if it was successfull. I added API Keys for all  Accounts so far. But yes, it seems some Values are off here. Some Accounts are complete fresh and the UI say they have 3 Mystic Coins, which should be not possible, because only the first Mystic Coin Daily Chest was opened, which contain only 2 Coins. Same for the Laurels, some Accounts say "2" which should be correct but two of the Accounts say "4", which is wrong. 

Gern geschehen, kein Problem ;)

**Comment by SamMackrill** [comment]
New version with fix for MC + Laurel counts : https://github.com/SamMackrill/gw2-alternator/releases/tag/V1.0.19
I will add your ideas as separate Issues, thanks!

**Comment by Caliban2017** [comment]
That sounds really good and i can confirm now - it worked! I had two Chest on the two Accounts i left untouched :) Looking forward for feature Versions and improvements.

One more little Bug, just a little: when i have already imported Accounts from Launch Buddy and i add again from Launch Buddy and have new Accounts in there, all old Accounts in the API Key List get duplicated. BUT after a restart of Alternator, the Duplicates are gone and everything is correct.

And one more Feature Request - remove unwanted imported Accounts from UI. Add additional Accounts would also be a nice Feature. At the Moment it is only possible to import Accounts, right?

**Comment by Caliban2017** [comment]
One more Feature Request: is it possible to change the Timestamp to the PCs local Timestamp? For me Alternator says Time is "00:45" but my real Time here is "2:45". So all last Login Timestamps are offset by 2 Hours for me.

**Comment by SamMackrill** [comment]
Please add any feature requests as a new issue, then they won't get lost :) (I have done this one)
The use of UTC is deliberate as that is the time the reset is measured in, but should be easy to add a flag and change the display times
